chia/cmds/param_types.py

```diff
@@ -0,0 +1,166 @@
+        if self.address_type is not address_type:
+            raise ValueError(f"Address must be of type {address_type}")
+        return self.original_address
+
+    def validate_address_type_get_ph(self, address_type: AddressType) -> bytes32:
+        if self.address_type is not address_type:
+            raise ValueError(f"Address must be of type {address_type}")
+        return self.puzzle_hash
+
+
+class AddressParamType(click.ParamType):
+    """
+    A Click parameter type for bech32m encoded addresses, it gives a class with the address type and puzzle hash.
+    """
+
+    name: str = "CliAddress"  # output type
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> CliAddress:
+        if isinstance(value, CliAddress):  # required by click
+            return value
+        try:
+            hrp, b32data = bech32_decode(value)
+            if hrp in ["xch", "txch"]:  # I hate having to load the config here
+                expected_prefix = ctx.obj.get("expected_prefix") if ctx else None
+                if expected_prefix is None:
+                    from chia.util.config import load_config
+                    from chia.util.default_root import DEFAULT_ROOT_PATH
+
+                    root_path = ctx.obj["root_path"] if ctx is not None else DEFAULT_ROOT_PATH
+                    config = load_config(root_path, "config.yaml")
+                    expected_prefix = selected_network_address_prefix(config)
+                    if ctx is not None:
+                        ctx.obj["expected_prefix"] = expected_prefix
+                # now that we have the expected prefix, we can validate the address
+                if hrp == expected_prefix:
+                    addr_type = AddressType.XCH
+                else:
+                    self.fail(f"Unexpected Address Prefix: {hrp}, are you sure its for the right network?", param, ctx)
+            else:
+                addr_type = AddressType(hrp)
+            return CliAddress(puzzle_hash=decode_puzzle_hash(value), address_type=addr_type, original_address=value)
+        except ValueError:
+            self.fail("Address must be a valid bech32m address", param, ctx)
+
+
+class Bytes32ParamType(click.ParamType):
+    """
+    A Click parameter type for bytes32 hex strings, with or without the 0x prefix.
+    """
+
+    name: str = "bytes32"  # output type
+
+    def convert(self, value: Any, param: Optional[click.Parameter], ctx: Optional[click.Context]) -> bytes32:
+        if isinstance(value, bytes32):  # required by click
+            return value
+        try:
+            return bytes32.from_hexstr(value)
+        except ValueError:
+            self.fail("Value must be a valid bytes32 hex string like a coin id or puzzle hash", param, ctx)
+
+
+# These are what we use in click decorators
+TRANSACTION_FEE = TransactionFeeParamType()
+AMOUNT_TYPE = AmountParamType()
+ADDRESS_TYPE = AddressParamType()
+BYTES32_TYPE = Bytes32ParamType()
```

Comment:
is there a convention to use all-caps type aliases for these? (other types usually use CamelCase)

Reply:
the click docs showed it using all caps, ill double check why

Reply:
I think I could use the python_case for them too

Reply:
There is a general Python convention, that I don't prefer, to `ALL_CAPS` 'constants'.  So I guess that's where they got it from?  https://click.palletsprojects.com/en/8.1.x/parameters/#implementing-custom-types  I'm not sure why having this reused instance is a good thing to begin with though.  Seems more like an extra layer that, as Arvid noted, looks a bit weird.  I guess it avoids folks forgetting to `()`.

Reply:
So what should I do

Reply:
We can have a group discussion about this, but it seems this topic remains open.

Reply:
This still looks untouched.  I don't necessarily have to get my way here but I think either a code change or another comment here is needed.

Reply:
removed as discussed